src/drivers/intel/hda/hda-dma.c

```diff
@@ -223,8 +223,6 @@ struct hda_chan_data {
 #endif
 };
 
-static int hda_dma_stop_common(struct dma_chan_data *channel);
```

Comment:
Not sure why this patch is needed if we are removing PAUSE.

Reply:
@lgirdwood I think to completely remove PAUSE support we have to do that in the driver (see `hda_link_dma_trigger()`:
https://github.com/thesofproject/linux/blob/ad489ae9054724c9b61e020fa8fefd84381a9490/sound/soc/sof/intel/hda-dai.c#L288-L289
). As long as the driver treats PAUSE and STOP differently, don't think the firmware can handle them equally.

Reply:
yes, I expect this to reduce code in core and drivers too.

Reply:
@lgirdwood sorry, if we **remove** PAUSE that's a complete different game. This PR **keeps** PAUSE and implements it with STOP. If we **remove** PAUSE that would mean, that we return an error in response, wouldn't it?

If we remove PAUSE, we should make this somehow that the CI diesn't fail, right? So, we should first remove pause-release tests from the CI, then modify the driver to cleanly return an error. Then add a test to CI to check that PAUSE and RELEASE return that error. Then remove support for it in the firmware

Reply:
> @lgirdwood sorry, if we **remove** PAUSE that's a complete different game. This PR **keeps** PAUSE and implements it with STOP. If we **remove** PAUSE that would mean, that we return an error in response, wouldn't it?

@lyakh we need to remove pause. If driver sends pause IPC then we treat as STOP and if we cannot do this (e.g. for the HDA DMA W/A then we return error). @ranj063 can help check the HDA DMA flow.

Reply:
@lgirdwood @lyakh The HDAudio flow is run -> pause -> reset. There is no stop. It's going to be very very hard to map the new concepts on hardware that assumes different steps.

Reply:
@plbossart this flow is fine and will remain. My main worry is that we had flows with both pause and stop combined to form a correct IP programming sequence.

Reply:
> Not sure why this patch is needed if we are removing PAUSE.

@lgirdwood yes, we remove the PAUSE trigger, i.e. `COMP_TRIGGER_PAUSE` - as long as `CONFIG_STOP_FOR_PAUSE` is enabled, which is the default.
But when the driver sends a PAUSE IPC upon a "pause" request from the user, we use `COMP_TRIGGER_STOP` in the SOF. Then, when a RELEASE IPC comes from the driver, we convert it to `COMP_TRIGGER_PRE_START`. But this alone doesn't work with HDA, because when a real STOP arrives, it doesn't yet stop the HDA DMA, and it is only stopped when `PCM_FREE` arrives. Since with PAUSE we don't have `PCM_FREE` we need this patch for HDA DMA.

Reply:
@lgirdwood we've discussed this some 2 years ago. Windows requires pause to be supported, so I really don't see where you are going with removing Pause. It's a non-starter.

Reply:
https://github.com/thesofproject/linux/issues/1339

Reply:
@plbossart Windows uses PAUSE IPC to STOP pipelines. It has no STOP named IPC.

e.g. the pipeline states.
```c
/**< Pipeline State */
enum ipc4_pipeline_state {
	/**< Invalid value */
	SOF_IPC4_PIPELINE_STATE_INVALID = 0,
	/**< Created but initialization incomplete */
	SOF_IPC4_PIPELINE_STATE_UNINITIALIZED = 1,
	/**< Resets pipeline */
	SOF_IPC4_PIPELINE_STATE_RESET = 2,
	/**< Pauses pipeline */
	SOF_IPC4_PIPELINE_STATE_PAUSED = 3,
	/**< Starts pipeline */
	SOF_IPC4_PIPELINE_STATE_RUNNING = 4,
	/**< Marks pipeline as expecting End Of Stream */
	SOF_IPC4_PIPELINE_STATE_EOS = 5,
	/**< Stopped on error */
	SOF_IPC4_PIPELINE_STATE_ERROR_STOP, <<<<<< this one is used to report a FW error to host.
	/**< Saved to the host memory */
	SOF_IPC4_PIPELINE_STATE_SAVED
};
```

and the SOF driver trigger logic 

```c
static int sof_ipc4_pcm_trigger(struct snd_soc_component *component,
				struct snd_pcm_substream *substream, int cmd)
{
	int state;

	/* determine the pipeline state */
	switch (cmd) {
	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
		state = SOF_IPC4_PIPE_PAUSED;
		break;
	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
	case SNDRV_PCM_TRIGGER_RESUME:
	case SNDRV_PCM_TRIGGER_START:
		state = SOF_IPC4_PIPE_RUNNING;
		break;
	case SNDRV_PCM_TRIGGER_SUSPEND:
	case SNDRV_PCM_TRIGGER_STOP:
		state = SOF_IPC4_PIPE_PAUSED;
		break;
	default:
		dev_err(component->dev, "%s: unhandled trigger cmd %d\n", __func__, cmd);
		return -EINVAL;
	}

	/* set the pipeline state */
	return sof_ipc4_trigger_pipelines(component, substream, state);
}
```

Reply:
It is a converged firmware requirement to support pause. You cannot remove this code now, or you will have to re-add it later.

It's also very useful at the kernel level to deal with all kinds of issues with suspend-resume.

I don't see any rationale for removing SNDRV_PCM_INFO_PAUSE. It has shown multiple issues in our state machines and forced us multiple times to do the right thing for suspend-resume.

Reply:
Sorry @lgirdwood the existing kernel sequences are part of the ABI. We don't control when kernel patches are integrated, which makes it impossible for the firmware to unilaterally decide when the changes are implemented. And since in most cases the firmware is updated much faster than the kernel you will most likely have to keep the sequences as is for quite a while.